src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -226,6 +252,201 @@ private void updateOriginalQueryResults(final CombineScoresDto combineScoresDTO,
+        List<CompoundTopDocs> queryTopDocs,
+        int indexOfFirstNonEmpty,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        boolean isKeyword = ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty)
+            .getTopDocs()
+            .getFirst()).collapseValues[0] instanceof BytesRef;
+
+        if (isKeyword) {
+            return collapseOnKeyword(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        } else {
+            return collapseOnNumeric(queryTopDocs, querySearchResults, sort, indexOfFirstNonEmpty, isFetchPhaseExecuted, combineScoresDTO);
+        }
+    }
+
+    private int collapseOnKeyword(
+        List<CompoundTopDocs> queryTopDocs,
+        List<QuerySearchResult> querySearchResults,
+        Sort sort,
+        int indexOfFirstNonEmpty,
+        final boolean isFetchPhaseExecuted,
+        CombineScoresDto combineScoresDTO
+    ) {
+        int totalScoreDocsCount = 0;
+        Map<BytesRef, FieldDoc> collapseValueToTopScoreDocMap = new HashMap<>();
+        Map<BytesRef, Integer> collapseValueToShardMap = new HashMap<>();
+        String collapseField = "";
+
+        HybridQueryFieldDocComparator comparator = new HybridQueryFieldDocComparator(
+            ((CollapseTopFieldDocs) queryTopDocs.get(indexOfFirstNonEmpty).getTopDocs().getFirst()).fields,
+            Comparator.comparing((ScoreDoc scoreDoc) -> scoreDoc.score)
+        );
+
+        for (int i = 0; i < querySearchResults.size(); i++) {
+            CompoundTopDocs updatedTopDocs = queryTopDocs.get(i);
+            List<ScoreDoc> updatedScoreDocs = updatedTopDocs.getScoreDocs();
+            if (updatedScoreDocs.isEmpty()) {
+                continue;
+            }
+            collapseField = ((CollapseTopFieldDocs) updatedTopDocs.getTopDocs().getFirst()).field;
+
+            for (ScoreDoc scoreDoc : updatedScoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+                BytesRef collapseValue = (BytesRef) fieldDoc.fields[fieldDoc.fields.length - 1];
+                if (collapseValue == null) {
+                    continue;
+                }
+
+                FieldDoc currentBestFieldDoc = collapseValueToTopScoreDocMap.get(collapseValue);
+                if (currentBestFieldDoc == null || comparator.compare(fieldDoc, currentBestFieldDoc) < 0) {
+                    collapseValueToTopScoreDocMap.put(BytesRef.deepCopyOf(collapseValue), fieldDoc);
+                    collapseValueToShardMap.put(BytesRef.deepCopyOf(collapseValue), i);
+                }
+            }
+        }
+
+        List<Map.Entry<BytesRef, FieldDoc>> entryList = new ArrayList<>(collapseValueToTopScoreDocMap.entrySet());
+
+        entryList.sort(Map.Entry.comparingByValue(comparator));
+
+        for (int index = 0; index < querySearchResults.size(); index++) {
```

Comment:
please pick more descriptive name for loop variable, maybe shardIndex?

Reply:
can we do a pre-filtering for performance

The current approach checks each entry against each shard index, which is O(n×m) complexity. Consider reorganizing the data structure to avoid this nested iteration, potentially using:

Map<Integer, List<Map.Entry<BytesRef, FieldDoc>>> shardToEntriesMap = entryList.stream()
    .collect(Collectors.groupingBy(entry -> collapseValueToShardMap.get(entry.getKey())));

// then for each index:
List<Map.Entry<BytesRef, FieldDoc>> relevantEntries = shardToEntriesMap.getOrDefault(index, Collections.emptyList());